types/lib/stdlib.go

```diff
@@ -0,0 +1,269 @@
+package lib
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	wire "github.com/cosmos/cosmos-sdk/wire"
+)
+
+// ListMapper is a Mapper interface that provides list-like functions
+// It panics when the element type cannot be (un/)marshalled by the codec
+
+type ListMapper interface {
+	// ListMapper dosen't checks index out of range
+	// The user should check Len() before doing any actions
+	Len(sdk.Context) int64
+	Get(sdk.Context, int64, interface{})
+	// Setting element out of range is harmful; use Push() when adding new elements
+	Set(sdk.Context, int64, interface{})
+	Delete(sdk.Context, int64)
+	Push(sdk.Context, interface{})
+	Iterate(sdk.Context, interface{}, func(sdk.Context, int64) bool)
+}
+
+type listMapper struct {
+	key    sdk.StoreKey
+	cdc    *wire.Codec
+	prefix string
+	lk     []byte
+}
+
+func NewListMapper(cdc *wire.Codec, key sdk.StoreKey, prefix string) ListMapper {
+	lk, err := cdc.MarshalBinary(int64(-1))
+	if err != nil {
+		panic(err)
+	}
+	return listMapper{
+		key:    key,
+		cdc:    cdc,
+		prefix: prefix,
+		lk:     lk,
+	}
+}
+
+func (lm listMapper) Len(ctx sdk.Context) int64 {
+	store := ctx.KVStore(lm.key)
+	bz := store.Get(lm.LengthKey())
+	if bz == nil {
+		zero, err := lm.cdc.MarshalBinary(0)
+		if err != nil {
+			panic(err)
+		}
+		store.Set(lm.LengthKey(), zero)
+		return 0
+	}
+	var res int64
+	if err := lm.cdc.UnmarshalBinary(bz, &res); err != nil {
+		panic(err)
+	}
+	return res
+}
+
+func (lm listMapper) Get(ctx sdk.Context, index int64, ptr interface{}) {
+	if index < 0 {
+		panic(fmt.Errorf("Invalid index in ListMapper.Get(ctx, %d, ptr)", index))
+	}
+	store := ctx.KVStore(lm.key)
+	bz := store.Get(lm.ElemKey(index))
+	if err := lm.cdc.UnmarshalBinary(bz, ptr); err != nil {
+		panic(err)
+	}
+}
+
+func (lm listMapper) Set(ctx sdk.Context, index int64, value interface{}) {
+	if index < 0 {
+		panic(fmt.Errorf("Invalid index in ListMapper.Set(ctx, %d, value)", index))
+	}
+	store := ctx.KVStore(lm.key)
+	bz, err := lm.cdc.MarshalBinary(value)
+	if err != nil {
+		panic(err)
+	}
+	store.Set(lm.ElemKey(index), bz)
+}
+
+func (lm listMapper) Delete(ctx sdk.Context, index int64) {
+	if index < 0 {
+		panic(fmt.Errorf("Invalid index in ListMapper.Delete(ctx, %d)", index))
+	}
+	store := ctx.KVStore(lm.key)
+	store.Delete(lm.ElemKey(index))
+}
+
+func (lm listMapper) Push(ctx sdk.Context, value interface{}) {
+	length := lm.Len(ctx)
+	lm.Set(ctx, length, value)
+
+	store := ctx.KVStore(lm.key)
+	store.Set(lm.LengthKey(), marshalInt64(lm.cdc, length+1))
+}
+
+func (lm listMapper) Iterate(ctx sdk.Context, ptr interface{}, fn func(sdk.Context, int64) bool) {
```

Comment:
Will fail if an element was deleted from the list (see above).

Reply:
Also, you can probably write this in a way that uses the builtin `Iterator` method of the `KVStore`, which would be more efficient and work correctly if elements are deleted.

Reply:
One thing that `Iterator` doesn't support is modifying value over the domain while there is an iterator for it(see: https://github.com/cosmos/cosmos-sdk/blob/master/types/store.go#L106). Maybe just returning `Iterator` of the `listMapper` is enough.